relay-event-normalization/src/normalize/nel.rs

```diff
@@ -59,3 +284,278 @@ pub fn create_log(nel: Annotated<NetworkReportRaw>, received_at: DateTime<Utc>)
+        assert_eq!(
+            extract_server_address("https://[2001:db8::1]:443/path"),
+            "[2001:db8::1]"
+        );
+        // Malformed IPv6 (missing closing bracket)
+        assert_eq!(
+            extract_server_address("http://[::1:8080/foo"),
+            "[::1:8080/foo"
+        );
+    }
+
+    #[test]
+    fn test_extract_server_address_ports() {
+        assert_eq!(
+            extract_server_address("https://example.com:443/path"),
+            "example.com"
+        );
+        assert_eq!(extract_server_address("http://localhost:3000"), "localhost");
+    }
+
+    #[test]
+    fn test_extract_server_address_edge_cases() {
+        // Empty URL part after protocol
+        assert_eq!(extract_server_address("https://"), "https://");
+        // No protocol
+        assert_eq!(extract_server_address("example.com"), "example.com");
+        // URL with fragments and queries
+        assert_eq!(
+            extract_server_address("https://example.com/path?query=value#fragment"),
+            "example.com"
+        );
+    }
+
+    #[test]
+    fn test_create_message() {
+        assert_eq!(
+            create_message("dns.unreachable", None),
+            "DNS server is unreachable"
+        );
+        assert_eq!(
+            create_message("http.error", Some(404)),
+            "The user agent successfully received a response, but it had a 404 status code"
+        );
+        assert_eq!(create_message("unknown_error", None), "unknown_error");
+    }
+
+    #[test]
+    fn test_create_log_basic() {
+        let received_at = Utc::now();
+
+        let body = BodyRaw {
+            ty: Annotated::new("http.error".to_owned()),
+            status_code: Annotated::new(500),
+            elapsed_time: Annotated::new(1000),
+            method: Annotated::new("GET".to_owned()),
+            protocol: Annotated::new("http/1.1".to_owned()),
+            server_ip: Annotated::new(IpAddr("192.168.1.1".to_owned())),
+            phase: Annotated::new(NetworkReportPhases::Application),
+            sampling_fraction: Annotated::new(1.0),
+            referrer: Annotated::new("https://example.com/referer".to_owned()),
+            ..Default::default()
+        };
+
+        let nel = NetworkReportRaw {
+            age: Annotated::new(5000),
+            ty: Annotated::new("network-error".to_owned()),
+            url: Annotated::new("https://example.com/api".to_owned()),
+            user_agent: Annotated::new("Mozilla/5.0".to_owned()),
+            body: Annotated::new(body),
+            ..Default::default()
+        };
+
+        let log = create_log(Annotated::new(nel), received_at).unwrap();
+
+        // Check basic fields
+        assert_eq!(log.level.into_value(), Some(OurLogLevel::Info));
+        assert_eq!(
+            log.body.into_value(),
+            Some(
+                "The user agent successfully received a response, but it had a 500 status code"
+                    .to_owned()
+            )
+        );
```

Comment:
Please when writing tests, try to orient yourself on how the rest of Relay writes test. We have specifically a dependency for tests like this called `insta`. 

Tests also need to be maintained and should still be reviewed and not blindly generated by AI.